components/merkledb/src/views/mod.rs

```diff
@@ -76,191 +81,80 @@ impl ChangeSet for ChangesRef<'_> {
     }
 }
 
-/// Base trait that allows to access and modify indexes.
-pub trait IndexAccess: Clone {
-    /// Type of the `changes` that will be applied to the database.
-    /// In case of `snapshot` changes is represented by the empty type,
-    /// because `snapshot` is read-only.
+/// Allows to read data from the database.
+///
+/// This trait is rarely needs to be used directly; [`Access`] is a more high-level trait
+/// encompassing access to database.
+///
+/// [`Access`]: trait.Access.html
+pub trait RawAccess: Clone {
+    /// Type of the `changes()` that will be applied to the database.
     type Changes: ChangeSet;
-    /// Reference to `Snapshot` used in `View` implementation.
+
+    /// Reference to a `Snapshot`.
     fn snapshot(&self) -> &dyn Snapshot;
-    /// Returns changes related to specific `address`.
+    /// Returns changes related to specific `address` compared to the `snapshot()`.
     fn changes(&self, address: &IndexAddress) -> Self::Changes;
 }
 
-/// Struct responsible for creating `view` and `state` for index with
-/// specified `address`. `View` contains changes, `state` contains
-/// metadata.
+/// Allows to mutate data in indexes.
 ///
-/// # Example
+/// This is a marker trait that is used as a bound for mutable operations on indexes.
+/// It can be used in the same way for high-level database objects:
 ///
 /// ```
-/// use exonum_merkledb::{Database, TemporaryDB, IndexBuilder, ListIndex};
+/// use exonum_merkledb::{access::{Access, RawAccessMut}, ListIndex, MapIndex};
 ///
-/// let db: Box<Database> = Box::new(TemporaryDB::new());
-/// let fork = db.fork();
-/// {
-///     let (view, _state) = IndexBuilder::new(&fork)
-///         .index_name("index")
-///         .build::<()>();
+/// pub struct Schema<T: Access> {
+///     list: ListIndex<T::Base, String>,
+///     map: MapIndex<T::Base, u64, u64>,
 /// }
 ///
+/// impl<T: Access> Schema<T>
+/// where
+///     T::Base: RawAccessMut,
+/// {
+///     pub fn mutate(&mut self) {
+///         self.list.push("foo".to_owned());
+///         self.map.put(&1, 2);
+///     }
+/// }
 /// ```
-#[derive(Debug)]
-pub struct IndexBuilder<T> {
-    index_access: T,
-    address: IndexAddress,
-    index_type: IndexType,
-}
+pub trait RawAccessMut: RawAccess {}
```

Comment:
Unlike the `RawAccess`, `RawAccessMut` is used almost everywhere, so the mutable/immutable code looks [asymmetric](https://github.com/exonum/exonum/pull/1523/files#diff-ee26286bb122f945e5926d1f6f710e23R18): for immutable access we use `Access`, but for mutable it's `RawAccessMut`. Maybe make this pair `Access`/`AccessMut`?

(or it even can be `MutAccess` to be closer to the plain english)

Reply:
`*Mut` suffix follows the naming conventions from std, such as `AsMut` and `BorrowMut`. Regarding `AccessMut`, I'm not opposed to the idea in principle, but `Access` / `RawAccessMut` are not *intended* to be symmetric: `RawAccessMut` is a mutable counterpart to `RawAccess` and thus concerns *indexes*, whereas `Access` concerns more high-level storage access.

Reply:
I understand that, but not sure if users should even think about it. IMHO for users MerkleDB should provide a set of simple abstractions that "just work", without need to understand inner mechanics.